### PR TITLE
add suport for masking

### PR DESCRIPTION
**Pull request by pixie79**
Adds support to bloblang for basic masking using either a fixed length mask, or partial left / right mask.
Allows to chose the masking character.

### CODE REVIEW

File: internal/impl/pure/bloblang_mask.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/redpanda-data/benthos/v4/internal/bloblang/query"
+	"github.com/redpanda-data/benthos/v4/public/bloblang"
+)
+
+func init() {
+	if err := bloblang.RegisterMethodV2("mask",
+		bloblang.NewPluginSpec().
+			Category(query.MethodCategoryParsing).
+			Description(`Masks a string using the given character, leaving X number of characters unmasked and returns a string.`).
+			Param(bloblang.NewInt64Param("mask_count").Description("the number of characters that will not be masked on the left or right hand side, in the case of a all mask, it is the number of mask characters to return giving a fixed length string, default is 0 which will return all characters masked.").Optional().Default(0)).
```
Comment (Jeffail) [inline internal/impl/pure/bloblang_mask.go:16]: I would name this just `count`.

File: internal/impl/pure/bloblang_mask.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+			Param(bloblang.NewStringParam("direction").Description("the direction to mask, left, right or all, default is all").Optional().Default("all")).
+			Param(bloblang.NewStringParam("mask_char").Description("the character used for masking, default is *").Optional().Default("*")).
```
Comment (Jeffail) [inline internal/impl/pure/bloblang_mask.go:18]: Same here, just `char`.

File: internal/impl/pure/bloblang_mask.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+			Example("", `root.values = this.body.mask(4)`,
+				[2]string{
+					`{"body":"noise=meow&animal=cat&fur=orange&fur=fluffy"}`,
+					`{"values":{"animal":"cat","fur":["orange","fluffy"],"noise":"meow"}}`,
```
Comment (Jeffail) [inline internal/impl/pure/bloblang_mask.go:22]: It looks as though this example is for a different method, and needs updating. But I do like the presence of `meow`, so maybe we can keep that somewhere in the new example.

File: internal/impl/pure/bloblang_mask.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	}
+}
+
+// maskString masks the string based on the given parameters.
+// `s` is the input string
+// `maskChar` is the character used for masking,
+// `direction` determines whether the start ("left") or the end ("right") of the string is masked or all for the whole string,
+// `maskCount` is the number of characters that will not be masked on the left or right hand side, in the case of a all mask, it is the number of mask characters to return giving a fixed length string.
+func maskString(s string, maskChar string, direction string, maskCount int) string {
+	println("using maskChar: ", maskChar)
```
Comment (Jeffail) [inline internal/impl/pure/bloblang_mask.go:50]:
```suggestion
```

File: internal/impl/pure/bloblang_mask.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+			Param(bloblang.NewStringParam("direction").Description("the direction to mask, left, right or all, default is all").Optional().Default("all")).
+			Param(bloblang.NewStringParam("mask_char").Description("the character used for masking, default is *").Optional().Default("*")).
+			Example("", `root.values = this.body.mask(4)`,
+				[2]string{
+					`{"body":"noise=meow&animal=cat&fur=orange&fur=fluffy"}`,
+					`{"values":{"animal":"cat","fur":["orange","fluffy"],"noise":"meow"}}`,
+				},
+			),
+		func(args *bloblang.ParsedParams) (bloblang.Method, error) {
+			maskCount, _ := args.GetOptionalInt64("mask_count")
```
Comment (Jeffail) [inline internal/impl/pure/bloblang_mask.go:26]: These errors ought to be checked, just a simple return on error is good enough.

**Review by Jeffail — CHANGES_REQUESTED** [review]
Hey @pixie79, implementation looks good overall, a few changes needed and then we're gravy.

File: internal/impl/pure/bloblang_mask.go (hunk 1 above)
Comment (pixie79) [reply internal/impl/pure/bloblang_mask.go]: Updated

File: internal/impl/pure/bloblang_mask.go (hunk 2 above)
Comment (pixie79) [reply internal/impl/pure/bloblang_mask.go]: Updated

File: internal/impl/pure/bloblang_mask.go (hunk 3 above)
Comment (pixie79) [reply internal/impl/pure/bloblang_mask.go]: lol yup forgot to update that one, done now

File: internal/impl/pure/bloblang_mask.go (hunk 4 above)
Comment (pixie79) [reply internal/impl/pure/bloblang_mask.go]: Removed the debug line

File: internal/impl/pure/bloblang_mask.go (hunk 5 above)
Comment (pixie79) [reply internal/impl/pure/bloblang_mask.go]: Added more error handling to return the errors

**Comment by pixie79** [comment]
@Jeffail I have made the updates but for some reason and I can't tell why my tests are now failing when I run:
go test internal/impl/pure/bloblang_mask_test.go

I note however that the tests for bloblang_string are also now failing for the same reason?
go test internal/impl/pure/bloblang_string_test.go

Has something else changed? - happy to fix if you can give me a clue

**Pull request closed without merging**